org.testeditor.web.dropwizard.xtext/src/main/java/org/testeditor/web/xtext/index/persistence/IndexUpdater.xtend

```diff
@@ -0,0 +1,100 @@
+package org.testeditor.web.xtext.index.persistence
+
+import com.google.common.annotations.VisibleForTesting
+import java.io.File
+import java.util.List
+import javax.inject.Inject
+import org.apache.commons.io.FilenameUtils
+import org.eclipse.emf.common.util.URI
+import org.eclipse.emf.ecore.resource.Resource
+import org.eclipse.jgit.diff.DiffEntry
+import org.eclipse.jgit.lib.Constants
+import org.slf4j.LoggerFactory
+import org.testeditor.web.xtext.index.XtextIndex
+
+class IndexUpdater {
+
+	static val logger = LoggerFactory.getLogger(IndexUpdater)
+
+	@Inject XtextIndex index
+
+	/**
+	 * Recursively traverses the file tree and adds all files to the
+	 * index that are relevant (with registered language extensions).
+	 */
+	def void addToIndex(File file) {
+		if (file.isDirectory && !shouldSkipDirectory(file)) {
+			file.listFiles.forEach[addToIndex]
+		} else if (file.isFile && isRelevantForIndex(file.path)) {
+			val uri = getAbsoluteFileURI(file)
+			index.updateOrAdd(uri)
+		}
+	}
+
+	@VisibleForTesting
+	protected def boolean shouldSkipDirectory(File directory) {
+		return directory.name == Constants.DOT_GIT
+	}
+
+	@VisibleForTesting
+	protected def boolean isRelevantForIndex(String filePath) {
+		if (filePath === null) {
+			return false
+		}
+		val knownLanguageExtensions = Resource.Factory.Registry.INSTANCE.extensionToFactoryMap.keySet
+		val fileExtension = FilenameUtils.getExtension(filePath).toLowerCase
+		return knownLanguageExtensions.exists[ext|ext.toLowerCase == fileExtension]
+	}
+
+	private def boolean isRelevantForIndex(DiffEntry diff) {
+		return isRelevantForIndex(diff.newPath) || isRelevantForIndex(diff.oldPath)
+	}
+
+	/**
+	 * Analyzes the Git diff and updates the index accordingly.
+	 */
+	def void updateIndex(File root, List<DiffEntry> diffs) {
+		for (diff : diffs) {
+			if (isRelevantForIndex(diff)) {
+				handleRelevantDiff(root, diff)
+			} else {
+				logger.debug("Skipping index update for irrelevant diff='{}'.", diff)
+			}
+		}
+	}
+
+	private def void handleRelevantDiff(File root, DiffEntry diff) {
+		logger.debug("Handling Git diff='{}'.", diff)
+		switch (diff.changeType) {
+			case ADD: {
```

Comment:
since oldPath or newPath can be relevant, we'll have to check here, wether the index action should really take place (e.g. rename hello.tcl -> hello.txt, remove hello.tcl but DO NOT ADD hello.txt)